Woocommerce revenue tracking improvements [MAILPOET-5485]

websupporter opened the pull request:
## Description

_N/A_

## Code review notes
Instead of hooking into special status changes, this PR suggest to hook into every status change. It extends the purchase entity and adds a `status` field. As soon as an order is created a purchase entity is created with the according status. When the status gets updated, the purchase entity gets updated as well.

We count as revenue only `completed`. So instead of storing just `completed` revenues, we filter the output for returning only `completed` purchase entities and counting them as "revenue". While this creates more data we store, it also opens up possibilities like tracking how many orders `failed` are `on-hold` etc. which might be interesting metrics as well.

## QA notes

We touched the revenue tracking code. Please verify order revenue is tracked and displayed in newsletter stats as expected. 
1) Create a newsletter and send it to a list
2) Make sure to click on links from a couple of received emails
3) Make a couple of orders and use emails from subscribers who clicked the links
4) Check the newsletter stats and observe that revenue was added
5) Try refunding an order and see change in stats

NOTE: For testing you will need premium from the branch same name (https://github.com/mailpoet/mailpoet-premium/pull/777)

## Linked PRs

https://github.com/mailpoet/mailpoet-premium/pull/777

## Linked tickets

[MAILPOET-5485]

## After-merge notes

_N/A_

## Tasks

- [x] I followed [best practices](https://codex.wordpress.org/I18n_for_WordPress_Developers) for translations
- [x] I added sufficient test coverage
- [x] I embraced TypeScript by either creating new files in TypeScript or converting existing JavaScript files when making changes


[MAILPOET-5485]: https://mailpoet.atlassian.net/browse/MAILPOET-5485?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

websupporter committed:
Extract purchase states to helper method

[MAILPOET-5485]

websupporter committed:
Listen to all status changes

This commit listens to all status changes but does not
alter the business logic as of yet.

[MAILPOET-5485]

websupporter committed:
Track only completed orders

[MAILPOET-5485]

websupporter committed:
Use getPurchaseStates helper method for past orders

[MAILPOET-5485]

websupporter committed:
Extract receiving the click from date to private method

[MAILPOET-5485]

websupporter committed:
Store current order status in revenue statistics

[MAILPOET-5485]

websupporter committed:
Track purchases regardless of status but return revenue status correctly

When we apply this commit every purchase will be tracked. When the shop manager updates an order
the status will be updated. For the statistics it will only return values for purchases with
the 'completed' status

[MAILPOET-5485]

websupporter committed:
Add test that revenue status is updated correctly

[MAILPOET-5485]

websupporter committed:
Test only real revenue is returned

[MAILPOET-5485]

websupporter committed:
Test tracker only returns revenues from completed orders

[MAILPOET-5485]

websupporter committed:
Add event option to tracking

[MAILPOET-5485]

websupporter committed:
Test event option is added correctly to the tracker

[MAILPOET-5485]

costasovo commented on mailpoet/lib/Migrations/Db/Migration_20230824_054259_Db.php:
```diff
@@ -0,0 +1,65 @@
+  ) {
+    parent::__construct($container);
+    $this->wooCommerceHelper = $container->get(Helper::class);
+  }
```
We should avoid using containers in DB-level migrations. WooCommerce helper doesn't work with Doctrine so it safe for now but may change.
I would suggest splitting the migration into two:
1) DB Level migration that creates the new column
2) App Level migration that fills the column.
The App-level migrations always run after the DB level so the column will be there and in the App-level migration, the container is available by default.

websupporter replied:
Ah! That makes sense. Didn't fully grasp the App/Db difference and how I should apply that. Thank you: 80c96c4

costasovo reviewed (changes requested):
@websupporter I like the idea of storing the state 👍 Thank you for thinking beyond the specification! Also thank you for the nice commits and improving test coverage! Great job!

When testing the tracking locally, I noticed that when I add a refund to an order and I don't change its status, the record in our table is not updated. It seems to be related to hooking only into status changes. Maybe we need to listen to general order updates or add listening for refunds.

websupporter committed:
Separate migration app and db logic

[MAILPOET-5485]

websupporter committed:
Extract tracking of existing statistic in own method

[MAILPOET-5485]

websupporter committed:
Test refunds are tracked

[MAILPOET-5485]

websupporter committed:
Track refunds

[MAILPOET-5485]

websupporter commented:
Thank you for your review @costasovo. Much appreciated! 

> When testing the tracking locally, I noticed that when I add a refund to an order and I don't change its status, the record in our table is not updated. It seems to be related to hooking only into status changes. Maybe we need to listen to general order updates or add listening for refunds.

It looks like you did a partial refund, which does not change the status. A fully refund should change the status:
https://github.com/woocommerce/woocommerce/blob/879136404d3b5445c1f1fb81c5e179d5333c96f4/plugins/woocommerce/includes/wc-order-functions.php#L650-L659

Nonetheless, we did not had a logic to adjust the revenue we generate from an order, when it got partially refunded, and actually not even when it got fully refunded as `$order->get_total()` will not change with a refund. `$order->get_remaining_refund_amount()` returns the amount you can refund, which is basically a calculation `$order->get_total() - $order->get_refunded_amount()`. I did not see any filters or some logic which could get into the way and think, if you really want to get how many revenue an order made, you need to use `get_remaining_refund_amount()`.

45fe6b3 adds a test and 66b666f passes this test.

costasovo reviewed (dismissed):
@websupporter Thanks for the fix and explanation! 👍 

When testing I missed that value didn't change even though I switched the status.

Aschepikov commented:
Hi @websupporter some checks have failed, can you take a look pls.

websupporter committed:
Switch double quotes with single quotes

[MAILPOET-5485]

websupporter commented:
@costasovo I needed to do a small code change. For some reason where `rev.status="unknown"` produces an SQL error. When I switch to `rev.status='unknown'` it works. So I added 9f15b4916182b209a88f3b3623bbf8657d0f3b2f and switched the quote types.

costasovo reviewed (dismissed):
@websupporter Thanks for the fix! The `'` vs `"` issue could be related to [ANSI_QUOTES](https://dev.mysql.com/doc/refman/8.0/en/sql-mode.html#sqlmode_ansi_quotes) mode.
[WPDB disables a bunch of incompatible modes](https://github.com/WordPress/WordPress/blob/21011a6ebb3080024d867dfdd6721b6e87fce6ae/wp-includes/class-wpdb.php#L645-L652) but it doesn't touch `ANSI_QUOTES`.
Good that you caught this issue. It could cause failing migrations for multiple users.
We must remember to use single quotes when working with WPDB. Could we catch this by a linter?

Aschepikov commented:
Hi @websupporter  I tested revenue stats it is working well, but I don't see my orders in the email stats. I tested on the 2 sites because I thought it could be connected to some site´s problem
![image](https://github.com/mailpoet/mailpoet/assets/113185437/d8ad0836-56db-4178-bd57-6380ae2f5131)
![image](https://github.com/mailpoet/mailpoet/assets/113185437/184e9594-5330-4ced-bed4-7fd8a57d8efb)
But on both sites, I see that the order is connected to the email but not in the email stats. 
![image](https://github.com/mailpoet/mailpoet/assets/113185437/77acf2ec-ccf2-471d-af30-5762f4405f59)
In this case, these orders should be in the email stats yes?

websupporter committed:
Apply revenue_status in subscriber stats

[MAILPOET-5485]

websupporter commented:
@Aschepikov good catch! Thank you!

Basically, it appeared to early in the subscriber statistics. It should appear, once the order has the status "completed". I updated the code.

Aschepikov commented:
@websupporter Thank you! Some checks are failing. I restarted it just in case they are flaky.

websupporter commented:
Thank you @Aschepikov. Build is green now.

Aschepikov commented:
fui on the second site, everything is ok

Aschepikov commented:
Oh on the second site, the completed status is working on the stats!

Aschepikov commented:
On this site completed status isn't tracked in the statistics, and it was completed before the last changes. I will retest everything on the 3 site to check that the old and new orders are displayed correctly 
![image](https://github.com/mailpoet/mailpoet/assets/113185437/4b9ddfcb-3756-4dea-9be0-4329daaee46b)

Aschepikov commented:
Also on the 1 site, new orders are tracking incorrectly. The order was made from the link in the 5123 email but is tracking as an order from another email. I will comment it here maybe it is connected with something else. I will try to reproduce it on another site
![image](https://github.com/mailpoet/mailpoet/assets/113185437/700b3466-8d41-46d1-a5f4-de879747d5c9)

Aschepikov commented:
Hi @websupporter can you take a look at this error from the third site? If it is not connected with the pr I will approve it as it worked okay on one of my sites.

The third site was broken after the installation of the pr. On the clean site, I installed the 4.26 prod version, prepared a few emails and orders.
On this site, the order is displayed incorrectly too, it should be connected just with the 5123 email 3.
![image](https://github.com/mailpoet/mailpoet/assets/113185437/1363eb3b-f1e1-426b-8b6f-d006575a856c)
After the installation of plugins prs, I have these errors. I had some errors on the previous sites but they were gone after the cache cleaning. 
![image](https://github.com/mailpoet/mailpoet/assets/113185437/28b68cd4-cb49-45f9-95c2-03bade188b5a)
![image](https://github.com/mailpoet/mailpoet/assets/113185437/63ca6354-fdb0-43a7-9131-0081fe693467)

Aschepikov reviewed (approved):
It will be tested one more time on a new site with the release build
